lex.go

```diff
@@ -226,3 +319,147 @@ func tokenize(reader io.Reader, tokenCh chan NgxToken) {
 
 	close(tokenCh)
 }
+
+type LuaLexer struct {
+	s *extScanner
+}
+
+func (ll *LuaLexer) Register(s *extScanner) []string {
+	ll.s = s
+	return []string{
+		"init_by_lua_block",
+		"init_worker_by_lua_block",
+		"exit_worker_by_lua_block",
+		"set_by_lua_block",
+		"content_by_lua_block",
+		"server_rewrite_by_lua_block",
+		"rewrite_by_lua_block",
+		"access_by_lua_block",
+		"header_filter_by_lua_block",
+		"body_filter_by_lua_block",
+		"log_by_lua_block",
+		"balancer_by_lua_block",
+		"ssl_client_hello_by_lua_block",
+		"ssl_certificate_by_lua_block",
+		"ssl_session_fetch_by_lua_block",
+		"ssl_session_store_by_lua_block",
+	}
+}
+
+//nolint:funlen,gocognit,gocyclo,nosec
+func (ll *LuaLexer) Lex(matchedToken string) <-chan NgxToken {
+	tokenCh := make(chan NgxToken)
+
+	tokenDepth := 0
+
+	go func() {
+		defer close(tokenCh)
+		var tok strings.Builder
+		var inQuotes bool
+
+		// special handling for'set_by_lua_block' directive
+		// ignore potential hardcoded credentials linter warning for "set_by_lua_block"
+		if matchedToken == "set_by_lua_block" /* #nosec G101 */ {
+			arg := ""
+			for {
+				if !ll.s.Scan() {
+					return
+				}
+				next := ll.s.Text()
+				if isSpace(next) {
+					if arg != "" {
+						tokenCh <- NgxToken{Value: arg, Line: ll.s.Line(), IsQuoted: false}
+						break
+					}
+
+					for isSpace(next) {
+						if !ll.s.Scan() {
+							return
+						}
+						next = ll.s.Text()
+					}
+				}
+				arg += next
+			}
+		}
+
+		// check that Lua block starts correctly
+		for {
+			if !ll.s.Scan() {
+				return
+			}
+			next := ll.s.Text()
+
+			if !isSpace(next) {
+				if next != "{" {
+					lineno := ll.s.Line()
+					tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: `expected "{" to start lua block`, Line: &lineno}}
+					return
+				}
+				tokenDepth++
+				break
+			}
+		}
+
+		// Grab everything in Lua block as a single token and watch for curly brace '{' in strings
+		for {
+			if !ll.s.Scan() {
+				return
+			}
+
+			next := ll.s.Text()
+			if err := ll.s.Err(); err != nil {
+				lineno := ll.s.Line()
+				tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: err.Error(), Line: &lineno}}
+			}
+
+			switch {
+			case next == "{" && !inQuotes:
+				tokenDepth++
+				if tokenDepth > 1 { // not the first open brace
```

Comment:
I'm a little confused why this check was need, but I think it's because `tok.WriteString` is repeated in every case vs putting the check for `tokenDepth == 0` after the `switch` statement and then calling `tok.WriteString`.

Reply:
I think it's just a minor difference, resolving.